internal/controller/mariadb_controller.go

```diff
@@ -341,6 +348,55 @@ func (r *MariaDBReconciler) reconcileStatefulSet(ctx context.Context, mariadb *m
 	return ctrl.Result{}, nil
 }
 
+func (r *MariaDBReconciler) reconcilePodLabels(ctx context.Context, mariadb *mariadbv1alpha1.MariaDB) (ctrl.Result, error) {
+	if !mariadb.Replication().Enabled {
+		return ctrl.Result{}, nil
+	}
+
+	podList := corev1.PodList{}
+	listOpts := &client.ListOptions{
+		LabelSelector: klabels.SelectorFromSet(
+			labels.NewLabelsBuilder().
+				WithMariaDBSelectorLabels(mariadb).
+				Build(),
+		),
+		Namespace: mariadb.GetNamespace(),
+	}
+	if err := r.List(ctx, &podList, listOpts); err != nil {
+		return ctrl.Result{}, fmt.Errorf("error listing Pods: %v", err)
+	}
+
+	for _, pod := range podList.Items {
+		var role = "replica"
+
+		if pod.Status.PodIP == "" || pod.Spec.NodeName == "" {
+			continue
+		}
+		podIndex, err := sts.PodIndex(pod.Name)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("error getting Pod '%s' index: %v", pod.Name, err)
+		}
+
+		if *podIndex == *mariadb.Status.CurrentPrimaryPodIndex {
```

Comment:
`mariadb.Status.CurrentPrimaryPodIndex` may not be set at this point under some conditions, potentially leading to dereferencing a nil pointer. Could we add a guard at the begining of this function to ensure we have `mariadb.Status.CurrentPrimaryPodIndex`  set ?